### PR TITLE
Pbench register triggers

### PR DESCRIPTION
**Pull request by zulcss**
Python pbench-register-tool-trigger

### CODE REVIEW

File: lib/pbench/cli/agent/commands/triggers/register.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            self.logger.error(
+                '%s: the stop trigger cannot have a colon in it: "%s"',
+                self.name,
+                self.context.stop,
+            )
+            return 1
+
+        # Remember this trigger
+        trigger = self.tool_group_dir / "__trigger__"
+        trigger.write_text("%s:%s\n" % (self.context.start, self.context.stop))
```
Comment (dbutenhof) [inline lib/pbench/cli/agent/commands/triggers/register.py:52]: Ah; should have looked at this first. So if a trigger file only has one line, the cat vs join in list-triggers doesn't matter. (Though it seems that joining lines may be a bit pointless in that case.)

**Review by dbutenhof — APPROVED** [review]

File: lib/pbench/cli/agent/commands/triggers/register.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+from pbench.cli.agent.options import common_options
+
+
+class TriggerRegister(TriggerCommand):
+    """Register tool trigger"""
+
+    def __init__(self, context):
+        super().__init__(context)
+
+        self.logger = setup_logging()
```
Comment (npalaska) [inline lib/pbench/cli/agent/commands/triggers/register.py:29]: `setup_logging` takes 2 arguments https://github.com/distributed-system-analysis/pbench/blob/master/lib/pbench/agent/utils.py#L5

File: lib/pbench/cli/agent/commands/triggers/list.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            if len(self.groups) == 0:
+                self.logger.error("%s: error fetching list of tool groups", self.name)
+                return 1
+
+            for group in self.groups:
+                trigger = self.gen_tools_group_dir(group) / "__trigger__"
+                if trigger.exists():
+                    with open(trigger) as f:
+                        data = f.read().splitlines()
+                    print("%s: %s" % (group, "  ".join(data)))
```
Comment (dbutenhof) [inline lib/pbench/cli/agent/commands/triggers/list.py:42]: I thought I saw this `splitlines`/`join` go away in the list-triggers PR? Now they're merged and it's back?

If `__trigger__` is a single line with `<start>:<stop>` then it probably doesn't matter. And if it somehow gets multiple lines your `" ".join` should give a similar effect to `$(cat ...)` but ... it still just feels funny. Mostly I guess I'm wondering if this was a deliberate reversion or an accidental regression on the merge.

Comment (zulcss) [reply lib/pbench/cli/agent/commands/triggers/list.py:42]: It needs to be rebased, just focus on the register.

**Review by portante — CHANGES_REQUESTED** [review]
Please rebase and get the unit tests working.  If you feel this is ready for review then, please take it out of `Draft` mode.  Thanks!

**Review by ndokos — DISMISSED** [review]
LGTM.

**Comment by ndokos** [comment]
You probably should take it out of Draft mode.

**Comment by zulcss** [comment]
#1884 needs to be merged first

**Review by dbutenhof — APPROVED** [review]

**Review by dbutenhof — CHANGES_REQUESTED** [review]
Same unit test failures here as in #1884 ... but with at least test-15 added.

File: agent/util-scripts/gold/pbench-register-tool-trigger/test-35.txt [hunk 4]
```diff
@@ -1,11 +1,16 @@
...
 /var/tmp/pbench-test-utils/pbench
+/var/tmp/pbench-test-utils/pbench/pbench.log
 /var/tmp/pbench-test-utils/pbench/tmp
 /var/tmp/pbench-test-utils/pbench/tools-v1-default
 /var/tmp/pbench-test-utils/pbench/tools-v1-default/testhost.example.com
 /var/tmp/pbench-test-utils/pbench/tools-v1-default/testhost.example.com/pidstat
 === /var/tmp/pbench-test-utils/pbench/tools-v1-default/testhost.example.com/pidstat:
 --- pbench tree state
++++ pbench.log file contents
+[ERROR][1900-01-01T00:00:00.000000] pbench-register-tool-trigger: the start trigger cannot have a colon in it: "START:DEFAULT"
```
Comment (dbutenhof) [inline agent/util-scripts/gold/pbench-register-tool-trigger/test-35.txt:15]: I hadn't thought this through reading the code; but this is actually a little weird, and I'm not sure we really need to be logging user input errors, especially at "ERROR" level.

**Review by dbutenhof — APPROVED** [review]

File: lib/pbench/cli/agent/commands/triggers/register.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            LOG.error(
+                '%s: the stop trigger cannot have a colon in it: "%s"',
+                self.name,
+                self.context.stop,
+            )
+            return 1
+
+        # Remember this trigger
+        trigger = self.tool_group_dir / "__trigger__"
+        trigger.write_text("%s:%s\n" % (self.context.start, self.context.stop))
```
Comment (portante) [inline lib/pbench/cli/agent/commands/triggers/register.py:56]: Can we change this to, `trigger.write_text(f"{self.context.start}:{self.context.stop}")`?

File: agent/Makefile [hunk 6]
```diff
@@ -256,6 +255,7 @@ install-python3-setup: install-util-scripts install-lib
 	(cd ..; SKIP_GENERATE_AUTHORS=1 SKIP_WRITE_GIT_CHANGELOG=1 python3 setup.py install --prefix=${DESTDIR}/python3)
 	${COPY} ${DESTDIR}/python3/bin/pbench-config ${UTILDIR}/
 	${COPY} ${DESTDIR}/python3/bin/pbench-clear-tools ${UTILDIR}/
+	${COPY} ${DESTDIR}/python3/bin/pbench-list-triggers ${UTILDIR}/
```
Comment (portante) [inline agent/Makefile:258]: We also need `pbench-register-tool-trigger`.

**Review by portante — CHANGES_REQUESTED** [review]

File: lib/pbench/test/functional/agent/cli/commands/triggers/test_register_tool_trigger.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
+import pytest
+
+
+def test_pbench_register_tool_trigger_help():
```
Comment (dbutenhof) [inline lib/pbench/test/functional/agent/cli/commands/triggers/test_register_tool_trigger.py:4]: Again, doc strings to describe each test case would be nice.

While combining multiple test cases into one method like your `test_register_triggers` is convenient to leverage common setup, you could also do that through a fixture and keep the test cases separate for easier reading and triage when something goes wrong.

**Review by dbutenhof — APPROVED** [review]

File: agent/Makefile [hunk 8]
```diff
@@ -255,6 +255,7 @@ install-python3-setup: install-util-scripts install-lib
 	${COPY} ${DESTDIR}/python3/bin/pbench-clear-results ${UTILDIR}/
 	${COPY} ${DESTDIR}/python3/bin/pbench-cleanup ${UTILDIR}/
 	${COPY} ${DESTDIR}/python3/bin/pbench-list-tools ${UTILDIR}/
+	${COPY} ${DESTDIR}/python3/bin/pbench-register-tool-triggers ${UTILDIR}/
```
Comment (portante) [inline agent/Makefile:258]: Don't we also have to remote it from [line 49](https://github.com/distributed-system-analysis/pbench/pull/1883/files#diff-34cc66de7395e97215a24040803ff3e7ab5dd07835a701dbb6a858012f696f7eL49)?

File: lib/pbench/test/functional/agent/cli/commands/triggers/test_register_tool_trigger.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        b"pbench-register-tool-trigger: the start trigger cannot have a colon in it:"
+        in err
+    )
+    assert exitcode == 1
+
+
+def test_register_triggers_invalid_stop(
+    monkeypatch, agent_config, pbench_run, pbench_cfg
+):
+    """pbench-register-tool-trigger with invalid start"""
```
Comment (portante) [inline lib/pbench/test/functional/agent/cli/commands/triggers/test_register_tool_trigger.py:58]: `s/invalid start/invalid stop/`

File: lib/pbench/test/functional/agent/cli/commands/triggers/test_register_tool_trigger.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        '--stop-trigger="STOP DEFAULT"',
+    ]
+    out, err, exitcode = pytest.helpers.capture(command)
+    assert exitcode == 0
+    trigger = pbench_run / "tools-v1-default" / "__trigger__"
+    assert trigger.exists()
+    assert b"tool trigger strings for start:" in out
+
+
+def test_register_triggers_invalid_start(
```
Comment (portante) [inline lib/pbench/test/functional/agent/cli/commands/triggers/test_register_tool_trigger.py:32]: Doc string from `*_stop()`?

File: lib/pbench/test/functional/agent/cli/commands/triggers/test_register_tool_trigger.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+
+def test_pbench_register_tool_trigger_help():
+    command = ["pbench-register-tool-trigger", "--help"]
+    out, err, exitcode = pytest.helpers.capture(command)
+    assert b"Usage: pbench-register-tool-trigger [OPTIONS]" in out
+    assert exitcode == 0
+
+
+def test_register_triggers(monkeypatch, agent_config, pbench_run, pbench_cfg):
```
Comment (portante) [inline lib/pbench/test/functional/agent/cli/commands/triggers/test_register_tool_trigger.py:11]: Perhaps a doc string like: "Verify triggers registered for default group".

File: lib/pbench/test/functional/agent/cli/commands/triggers/test_register_tool_trigger.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        "pbench-register-tool-trigger",
+        "--group=default",
+        '--start-trigger="START DEFAULT"',
+        '--stop-trigger="STOP DEFAULT"',
+    ]
+    out, err, exitcode = pytest.helpers.capture(command)
+    assert exitcode == 0
+    trigger = pbench_run / "tools-v1-default" / "__trigger__"
+    assert trigger.exists()
+    assert b"tool trigger strings for start:" in out
```
Comment (portante) [inline lib/pbench/test/functional/agent/cli/commands/triggers/test_register_tool_trigger.py:29]: Can we read the contents of `__trigger__` and verify we see the expected values?

**Review by portante — CHANGES_REQUESTED** [review]

**Review by portante — APPROVED** [review]

File: agent/Makefile (hunk 8 above)
Comment (portante) [reply agent/Makefile:258]: This still needs to be addressed.

**Review by portante — CHANGES_REQUESTED** [review]
Actually, one more minor change.

**Review by portante — APPROVED** [review]

**Review by ndokos — APPROVED** [review]

**Comment by ndokos** [comment]
LGTM, but needs a rebase.

**Review by dbutenhof — APPROVED** [review]

**Review by npalaska — DISMISSED** [review]

**Review by portante — APPROVED** [review]

**Review by dbutenhof — APPROVED** [review]

**Pull request merged**